File: scripts/_cli/explain_last/inputs.py
"""Resolve the ``inputs`` and ``pack`` why-slots for the trace.

Reuses :mod:`scripts.config.profiles` and :mod:`scripts.config.presets`
so the rendered chain matches what the runtime loader actually
consulted (no parallel logic — the v1 ``explain config`` surface
already covers this and we read through the same resolvers).
"""
from __future__ import annotations

import logging
from contextlib import contextmanager
from pathlib import Path
from typing import Any

from scripts._cli.explain_last.scrubber import scrub_string
from scripts._lib.agent_settings import DEFAULT_PROJECT_FILE, load_agent_settings
from scripts.config import presets, profiles

_DEFAULT_COST_PROFILE = "balanced"
_SILENCED_LOGGERS = ("scripts.config.profiles", "scripts.config.presets")


@contextmanager
def _silence_resolver_warnings():
    # Profile / preset loggers emit absolute paths on read failure;
    # the explain trace must never leak them. Failures bubble up via
    # ProfileError / PresetError anyway, so the warning has no signal.
    previous: list[tuple[logging.Logger, int]] = []
    for name in _SILENCED_LOGGERS:
        log = logging.getLogger(name)
        previous.append((log, log.level))
        log.setLevel(logging.ERROR)
    try:
        yield
    finally:
        for log, level in previous:
            log.setLevel(level)


def _load_settings(project_root: Path) -> dict[str, Any]:
    path = project_root / DEFAULT_PROJECT_FILE
    if not path.exists():
        return {}
    return load_agent_settings(project_path=path) or {}


def build(project_root: Path) -> dict[str, Any] | None:
    """Return the ``inputs`` slot or ``None`` on broken settings.

    Each knob carries its source (``pack | profile | preset | user |
    env | runtime | default``) so the renderer's table can attribute
    the value to the layer that won the resolution race.
    """
    try:
        with _silence_resolver_warnings():
            settings = _load_settings(project_root)
            resolved_profile = profiles.resolve_profile(
                project_root=project_root,
                user_settings=settings,
            )
            resolved_preset = presets.resolve_preset(
                project_root=project_root,
                user_settings=settings,
                profile_preset_id=resolved_profile.preset_id,
            )
    except (profiles.ProfileError, presets.PresetError, OSError):
        return None
    cost_profile = settings.get("cost_profile") if isinstance(settings, dict) else None
    cost_profile_source = "user" if cost_profile else "default"
    if not cost_profile or cost_profile == "__COST_PROFILE__":
        cost_profile = _DEFAULT_COST_PROFILE
        cost_profile_source = "default"
    return {
        "profile": scrub_string(resolved_profile.id),
        "preset": scrub_string(resolved_preset.id),
        "cost_profile": scrub_string(str(cost_profile)),
        "source_per_knob": {
            "profile": resolved_profile.source,
            "preset": resolved_preset.source,
            "cost_profile": cost_profile_source,
        },
    }


def _pack_marker(project_root: Path) -> Path | None:
    for candidate in (
        project_root / ".agent-pack.yml",
        project_root / ".agent-src.uncompressed" / ".agent-pack.yml",
    ):
        if candidate.exists():
            return candidate
    return None


def build_pack(project_root: Path) -> dict[str, Any] | None:
    """Return the active workspace pack or ``None``.

    The discovery pipeline (R3) writes a ``.agent-pack.yml`` marker
    when a pack is bound to the project. We surface only the id and a
    short reason; pack-internal payloads stay invisible to the trace.
    """
    marker = _pack_marker(project_root)
    if marker is None:
        return None
    try:
        import yaml  # type: ignore[import-not-found]
    except ImportError:
        return None
    try:
        raw = yaml.safe_load(marker.read_text(encoding="utf-8")) or {}
    except (OSError, yaml.YAMLError):
        return None
    if not isinstance(raw, dict):
        return None
    pack_id = raw.get("id") or raw.get("pack")
    if not isinstance(pack_id, str) or not pack_id.strip():
        return None
    reason = raw.get("reason") or f"declared in {marker.name}"
    return {
        "id": scrub_string(pack_id.strip()),
        "reason": scrub_string(str(reason)),
    }


__all__ = ["build", "build_pack"]

File: src/scripts/_cli/explain_last/inputs.ts
/**
 * Resolve the `inputs` and `pack` why-slots for the trace.
 *
 * TypeScript twin of `src/scripts/_cli/explain_last/inputs.py` (ADR-200).
 * Behaviour mirrors the Python original EXACTLY — same resolver reuse,
 * same `None`-on-error branches, same per-knob source attribution, same
 * pack-marker discovery order and key shape. No behaviour changes.
 *
 * Reuses `scripts.config.profiles` and `scripts.config.presets` so the
 * rendered chain matches what the runtime loader actually consulted (no
 * parallel logic — the v1 `explain config` surface already covers this
 * and we read through the same resolvers).
 */
import { createRequire } from 'node:module';
import * as fs from 'node:fs';
import * as path from 'node:path';

import { DEFAULT_PROJECT_FILE, load_agent_settings, logger } from '../../_lib/agent_settings.js';
import * as presets from '../../config/presets.js';
import * as profiles from '../../config/profiles.js';
import { scrub_string } from './scrubber.js';

const _require = createRequire(import.meta.url);

const _DEFAULT_RULE_LOADING_TIER = 'balanced';
const _SILENCED_LOGGERS: readonly string[] = [
    'scripts.config.profiles',
    'scripts.config.presets',
];

/**
 * Faithful twin of the Python `_silence_resolver_warnings` context manager.
 *
 * The Python original raises the level of the `scripts.config.profiles` /
 * `scripts.config.presets` loggers to ERROR so a profile/preset read
 * failure does not leak an absolute path on stderr. The TypeScript config
 * twins funnel through a single shared `agent_settings.logger` that only
 * appends to `logger.records` (no stderr emission, no per-name level
 * threshold), so there is nothing to silence — the warning never reaches
 * a user-visible channel. This wrapper therefore runs the body unchanged;
 * it preserves the structure (and the `_SILENCED_LOGGERS` name list) for
 * fidelity and never alters the stdout the trace is built from.
 * (ADR-200 documented divergence — logging-model difference, stdout
 * byte-parity unaffected.)
 */
function _silence_resolver_warnings<T>(body: () => T): T {
    // Touch the names + logger so the structural intent is visible to a
    // reader and the import is exercised; behaviour is a pass-through.
    void _SILENCED_LOGGERS;
    void logger;
    return body();
}

function _load_settings(project_root: string): Record<string, unknown> {
    const p = path.join(project_root, DEFAULT_PROJECT_FILE);
    if (!fs.existsSync(p)) {
        return {};
    }
    return (load_agent_settings({ project_path: p }) as Record<string, unknown>) ?? {};
}

/** Python truthiness for the `or` chains below (empty string / 0 / null falsy). */
function _pyTruthy(value: unknown): boolean {
    if (value === null || value === undefined) {
        return false;
    }
    if (typeof value === 'string') {
        return value.length > 0;
    }
    if (typeof value === 'number') {
        return value !== 0;
    }
    if (typeof value === 'boolean') {
        return value;
    }
    if (Array.isArray(value)) {
        return value.length > 0;
    }
    if (typeof value === 'object') {
        return Object.keys(value as object).length > 0;
    }
    return true;
}

export function build(project_root: string): Record<string, unknown> | null {
    let resolved_profile: profiles.ResolvedProfile;
    let resolved_preset: presets.ResolvedPreset;
    let settings: Record<string, unknown>;
    try {
        const result = _silence_resolver_warnings(() => {
            const s = _load_settings(project_root);
            const rp = profiles.resolve_profile({
                project_root,
                user_settings: s,
            });
            const rpre = presets.resolve_preset({
                project_root,
                user_settings: s,
                profile_preset_id: rp.preset_id,
            });
            return { settings: s, resolved_profile: rp, resolved_preset: rpre };
        });
        settings = result.settings;
        resolved_profile = result.resolved_profile;
        resolved_preset = result.resolved_preset;
    } catch (exc) {
        // (profiles.ProfileError, presets.PresetError, OSError) → None.
        if (
            exc instanceof profiles.ProfileError
            || exc instanceof presets.PresetError
            || _isOsError(exc)
        ) {
            return null;
        }
        throw exc;
    }
    // Mirror Python:
    //   rule_loading_tier = (
    //       (settings.get("rule_loading_tier") or settings.get("cost_profile"))
    //       if isinstance(settings, dict) else None
    //   )
    // `.get(...)` is `null` when absent; `or` falls through on a falsy first.
    const isDict = typeof settings === 'object' && settings !== null;
    let rule_loading_tier: unknown;
    if (isDict) {
        const primary = settings.rule_loading_tier ?? null;
        rule_loading_tier = _pyTruthy(primary)
            ? primary
            : (settings.cost_profile ?? null);
    } else {
        rule_loading_tier = null;
    }
    let rule_loading_tier_source = _pyTruthy(rule_loading_tier) ? 'user' : 'default';
    if (!_pyTruthy(rule_loading_tier) || rule_loading_tier === '__RULE_LOADING_TIER__') {
        rule_loading_tier = _DEFAULT_RULE_LOADING_TIER;
        rule_loading_tier_source = 'default';
    }
    return {
        profile: scrub_string(resolved_profile.id),
        preset: scrub_string(resolved_preset.id),
        rule_loading_tier: scrub_string(String(rule_loading_tier)),
        source_per_knob: {
            profile: resolved_profile.source,
            preset: resolved_preset.source,
            rule_loading_tier: rule_loading_tier_source,
        },
    };
}

function _pack_marker(project_root: string): string | null {
    for (const candidate of [
        path.join(project_root, '.agent-pack.yml'),
        path.join(project_root, '.agent-src.uncondensed', '.agent-pack.yml'),
    ]) {
        if (fs.existsSync(candidate)) {
            return candidate;
        }
    }
    return null;
}

export function build_pack(project_root: string): Record<string, unknown> | null {
    const marker = _pack_marker(project_root);
    if (marker === null) {
        return null;
    }
    let YAML: typeof import('yaml');
    try {
        // Lazy require mirrors Python's lazy `import yaml`.
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        YAML = _require('yaml') as typeof import('yaml');
    } catch {
        // ImportError → None.
        return null;
    }
    let raw: unknown;
    try {
        const text = fs.readFileSync(marker, 'utf-8');
        raw = YAML.parse(text, { version: '1.1' });
        if (raw === null || raw === undefined) {
            raw = {};
        }
    } catch {
        // (OSError, yaml.YAMLError) → None.
        return null;
    }
    if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
        return null;
    }
    const rawDict = raw as Record<string, unknown>;
    const pack_id = _pyTruthy(rawDict.id) ? rawDict.id : (rawDict.pack ?? null);
    if (typeof pack_id !== 'string' || pack_id.trim() === '') {
        return null;
    }
    const reasonRaw = _pyTruthy(rawDict.reason)
        ? rawDict.reason
        : `declared in ${path.basename(marker)}`;
    return {
        id: scrub_string(pack_id.trim()),
        reason: scrub_string(String(reasonRaw)),
    };
}

/** Heuristic for the Python `except ... OSError` arm. */
function _isOsError(exc: unknown): boolean {
    return (
        exc instanceof Error
        && typeof (exc as NodeJS.ErrnoException).code === 'string'
    );
}
